src/full/Agda/TypeChecking/Serialise/Base.hs

```diff
@@ -167,43 +245,52 @@ type S a = ReaderT Dict IO a
 -- 'TCM' is not used because the associated overheads would make
 -- decoding slower.
 
-type R a = ExceptT TypeError (StateT St IO) a
+type R = StateT St IO
 
 -- | Throws an error which is suitable when the data stream is
 -- malformed.
 
 malformed :: R a
-malformed = throwError $ GenericError "Malformed input."
+malformed = liftIO $ E.throwIO $ E.ErrorCall "Malformed input."
+{-# NOINLINE malformed #-} -- 2023-10-2 András: cold code, so should be out-of-line.
 
 class Typeable a => EmbPrj a where
   icode :: a -> S Int32  -- ^ Serialization (wrapper).
   icod_ :: a -> S Int32  -- ^ Serialization (worker).
   value :: Int32 -> R a  -- ^ Deserialization.
 
   icode a = do
+    !r <- icod_ a
     tickICode a
-    icod_ a
+    pure r
+  {-# INLINE icode #-}
 
   -- Simple enumeration types can be (de)serialized using (from/to)Enum.
-
   default value :: (Enum a, Bounded a) => Int32 -> R a
   value i =
     let i' = fromIntegral i in
     if i' >= fromEnum (minBound :: a) && i' <= fromEnum (maxBound :: a)
-    then pure (toEnum i')
+    then pure $! toEnum i'
     else malformed
 
   default icod_ :: (Enum a, Bounded a) => a -> S Int32
-  icod_ = return . fromIntegral . fromEnum
+  icod_ x = return $! fromIntegral $! fromEnum x
```

Comment:
Ah, this line is what I was looking for previously but couldn't find it any more.
This would be an opportunity to use a strict composition `(.!)` which we could place in `Agda.Utils.Function`.